[FEAT/#10] 번호 인증 검사 기능 구현

yummygyudon opened the pull request:
<!--
name: Makers Auth Feature PR template
about: 구현한 기능과 변경 사항에 대해 구체적으로 작성해주세요
title: '[FEAT/{#이슈번호}] 기능 내용'
labels: ''
assignees: ''
-->

<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## Related Issue 🚀
- closes #10 
- closes #3 

<br/>

## Work Description ✏️
- 번호 인증 검사 기능을 구현했습니다.

(Jpa 테스트의 경우, 리뷰를 받는 과정에서 추가 진행할 예정입니다.)

<br/>

## PR Point 📸
<!-- 피드백을 받고 싶은 부분을 적어주세요. -->

### ※ Verify(검증) 메서드 역할 분리

- 현재 구조
  - `verify()` 메서드 구현체 내에서 데이터 삭제까지 진행
  - 요청 데이터 - 조회 데이터가 동일할 경우, 조회 데이터에 대해 삭제하는 로직이 한 메서드에 포함되어 있습니다.
- 고민 구조
  1.  Usecase 내 `delete()` 메서드 추가 구현하여 함수 분리
  2. (현재 구조 유지) 

@sung-silver  @hyunw9 
위 사항에 대한 여러분들의 의견이 궁금합니다!! 가감없이 의견 공유 주세요!!

<br/>

### ※ Usercase 메서드 반환 타입
본 PR에서는 구현한 `verify()` 메서드의 반환 타입으로 `AuthResponse.VerifyResult` DTO 객체를 반환하여
해당 반환 객체를 Response Body Data로 주입하도록 구현했습니다.

본래 계층 간 협력 관계에서 협력 메시지 객체를 사용하도록 하여 도메인 객체를 외부로부터 보존하기 위해 위 방법을 채택했습니다!

이 때, 아래와 같은 상황이 우려되었습니다.
" **_한 개의 API 메서드 내부에서 복수의 Usecase 메서드를 호출하는 경우_** "

<br/>

Usecase의 메서드 반환 타입이 모두 Response Body DTO 객체라면
복수의 Usecase 메서드를 호출했을 때, " 복수의 Response Body DTO를 재조합하여 새로운 Response Body DTO 객체를 생성 "해야하는 상황이 발생할 것으로 예상되었습니다.

@sung-silver  @hyunw9 
이러한 상황은 바람직한 구조는 아니라고 판단되어 여러분들과 논의를 해보면 좋을 것 같아 가져오게 되었습니다!!
편하게 의견 공유 주세요!!

yummygyudon committed:
[FEAT] 번호 인증 검사 API Response Body 데이터 정의

yummygyudon committed:
[FEAT] 번호 인증 Usecase 및 Command 정의

Return 타입으로 Response DTO를 선언했습니다.

yummygyudon committed:
[FEAT] 번호 인증 데이터 Out Port 정의 및 JPA 메서드 선언

yummygyudon committed:
[FEAT] 번호 인증 검사 In Port Usecase 구현

요청 데이터 기반 도메인 객체와 조회 데이터 기반 도메인 객체의 동일여부에 따라 검증 여부를 판별할 수 있도록 합니다.
- PhoneVerification 객체에 대해 `@EqualsAndHashCode` 어노테이션을 추가하여 필드값을 통한 비교가 가능하도록 했습니다.

* AuthRequest 객체에서 Command로 변환하는 메서드도 정의했습니다.
* 누락된 Repository 내 인증 이력 조회가 없을 경우에 대한 실패 코드 정의 작업 내용을 추가했습니다.

yummygyudon committed:
[FEAT] 번호 인증 요청에 대한 API 메서드 구현

yummygyudon committed:
[TEST] PhoneVerification 객체 간 동일여부 판단 메서드 테스트

yummygyudon committed:
[FIX] PhoneVerification Inner Record `@EqualsAndHashCode` 추가

- 내부 `PhoneVerification.VerificationCode` Record 또한 객체로서 생성되어 HashCode 값이 상이하기 때문에 발생한 테스트 불통과 문제를 해결했습니다.

yummygyudon committed:
[REFACTOR] Transaction Helper 객체 구현

기존 interface의 default 접근제어자 메서드들에 접근 가능한 환경을 개선했습니다.
- 구현 클래스가 많아지기 때문에 내부적으로 도메인별 패키징을 추가했습니다.

yummygyudon committed:
[REFACTOR] PhoneVerification Out Port 메서드명 개선

Transaction의 성격이 강한 메서드 명보다 직관적인 메서드 명으로 변경했습니다.

yummygyudon committed:
[CHORE] 소스코드 변경 사항 및 패키징 추가 반영

yummygyudon committed:
[CHORE] 와일드 카드 제거 및 Usecase 반환 데이터 변경 반영

hyunw9 commented:
코드 리뷰는 오프라인으로 진행했기 때문에 생략하겠슴다~~~

## 1.
제 생각에는요 `useCase` 자체가 행위 기반으로 이루어진 집합이기 때문에, `delete`와 같은 추가적인 메서드를 사용하는 것 보다
현재의 방식이 더 적합하다고 생각합니다 ! 

다만 verify 메서드에서 두 가지의 역할을 하고 있음이 분명하기 때문에, SRP를 유지가 목적이라면 Event 발행 구조를 적용해도 될 것 같지만
이벤트 발행을 해도, 어차피 한 트랜잭션에 묶여서 처리되기 때문에 가독성 향상의 이점 빼고는 크게 얻어가는 부분이 없다고 생각합니다.

따라서 제 의견은 현행 유지. 입니다.

## 2.

만약 한 API 엔드포인트에서 여러 유스케이스를 호출하는 구조라면, 유즈케이스 설계가 잘못된 경우라고 생각합니다! 
저는 VO 객체로 반환하는 편이 더 낫다고 생각합니다.

hyunw9 reviewed (approved):
고생했슴둥~~

sung-silver commented on src/main/java/sopt/makers/authentication/application/auth/dto/response/AuthResponse.java:
```diff
@@ -2,7 +2,12 @@
-public final class AuthResponse {}
+public final class AuthResponse {
+
+  public record VerifyResult(@JsonProperty("isVerified") boolean isVerified) {}
```
`p3`

회원가입에서 번호 인증을 사용하게 되면 운영진이 제공해준 사용자 정보를 식별할 수 있는 값도 전달되어야 할 것 같은데 어떻게 생각하시나요? 컨트롤러를 분리하는게 좋을지 response에서 필드를 추가해 재사용하는게 좋을지 고민되어 일단 response에 남겨봅니다!

yummygyudon replied:
오호옷!!

식별 데이터라 하심은
임원진으로부터 전달받은 "**신규회원 csv를 저장한 테이블**"에서의 ID를 넘겨주자는 거죠?!

전 좋은 것 같습니다!!
그래야만 소셜 인증 요청 시, 인증된 csv 내 신규회원 식별 값을 함께 넘겨 해당 DB에서 삭제까지 용이하게 할 수 있겠군요 
(저희가 논의했을 땐, 소셜인증까지 완료했을 때만 csv 저장 데이터를 삭제하자고 했었으니까!!)

sung-silver replied:
넵 근데 고려되어야 할 점이 회원가입 시점에서는 csv 파일을 저장한 id를 넘겨주셔야하지만 소셜계정 변경을 위해 번호인증을 한 경우에는 실제 유저의 id를 넘겨주시거나 다른 값(현재 사용되고 있는 socialPlaltformId 등)을 전달 해야할 것 같은데 어떻게 생각하시나요?

yummygyudon replied:
- 논의 결과 : 상황에 따라 API의 역할에 대해 명확히 분리했습니다 (번호 인증 != 유저 인증)
- 반영 결과 : 번호 인증 성공 여부만 반환합니다.

sung-silver commented on src/main/java/sopt/makers/authentication/database/PhoneVerificationRepositoryImpl.java:
```diff
@@ -0,0 +1,38 @@
+package sopt.makers.authentication.database;
```
`p3`

파일의 위치가 여기 맞을까요??

yummygyudon replied:
넵!! 이건 어제 현장에서 구두 설명 드렸으니 Resolve 처리하도록 하겠습니다!

sung-silver commented on src/main/java/sopt/makers/authentication/database/PhoneVerificationRepositoryImpl.java:
```diff
@@ -0,0 +1,38 @@
+@RequiredArgsConstructor
+public class PhoneVerificationRepositoryImpl implements PhoneVerificationRepository {
+
+  private final PhoneVerificationRegister register;
+  private final PhoneVerificationRetriever retriever;
+  private final PhoneVerificationRemover remover;
```
좋다...

sung-silver commented on src/main/java/sopt/makers/authentication/database/rdb/repository/auth/PhoneVerificationJpaRepository.java:
```diff
@@ -0,0 +1,39 @@
+                verification.getVerificationType())
+            .orElseThrow(() -> new AuthException(AuthFailure.NOT_FOUND_PHONE_VERIFICATION));
+    return targetPhoneVerificationEntity;
+  }
```
`p4`

Remover에서 orElseThrow()에 대한 처리를 하지 않고 repository에서 처리한 이유가 궁금합니다! (단순 궁금)

yummygyudon replied:
Retriever 말씀하신거겠죠?!
확실히 여기보단 Helper 클래스에 있는 것이 좋겠군요

피드백 감사드립니다!!

sung-silver replied:
웹으로 보다보니 왔다갔다를 잘못한..것 같네요 ㅎㅎ 맞습니다

yummygyudon committed:
[FIX] 요구사항에 따른 PhoneVerification Null 가능 반영

yummygyudon committed:
[TEST] PhoneVerification Null 가능 반영 정상 작동 여부 확인

sung-silver commented on src/main/java/sopt/makers/authentication/usecase/auth/service/VerifyVerificationService.java:
```diff
@@ -0,0 +1,30 @@
+    boolean isVerified = targetVerification.equals(findVerification);
+
+    if (isVerified) {
+      phoneVerificationRepository.deletedByPhoneVerification(findVerification);
```
`p5`
여기는 개행 없어도 될 것 같습니다!

yummygyudon replied:
확인이 완료된 "인증 이력 삭제"에 대한 행위를 가시적으로 분리하고 싶어서 개행했는데...!!
개행이 없어도 될 것 같다고 판단하신 이유를 조심스레 여쭤봅니다!!!

sung-silver replied:
지금 다시 보니까 변수 선언 이후에 설명문이 오고있어서 개행이 오는게 맞는 것 같은데 아마 카페에서 스플릿뷰로 보면서 비교하다보니 2줄로 착각했던 것 같습니다! 지금 전체화면으로 보니까 잘 되어있네요!! 혼란을 드려 죄송합니다,,

변수선언
한줄띄고
나머지 실행문

이 구조가 맞습니다!

sung-silver reviewed (changes requested):
궁금한 점 남겨보았습니다!!

1. verify 메서드 역할 분리
- 가독성을 생각하면 delete()로 분리해도 좋을 것 같습니다!
- delete로 분리하더라도 verify내에서 호출되면 두 가지의 일을 하는건 동일하다는 느낌? ㅎ

2. 한 개의 API 메서드 내부에서 복수의 Usecase 메서드를 호출하는 경우
- Usecase -> 서비스의 인터페이스인데 두 개의 dto를 조합하여 데이터를 만들어내는 경우는 없지 않을지..! 조심스레 걱정 안해도 될 것 같다고 생각합니다!

yummygyudon committed:
[REFACTOR] 트랜젝션 처리/예외 처리 계층 변경

helper 레이어를 추가함으로서 기존 Repository 인터페이스 내에서 default 메서드 정의로 예외 처리하던 구조를 helper 객체로 이관했습니다.
- Repository 객체는 실제 쿼리 실행 역할만 수행할 수 있도록 개선했습니다.
- [리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/15#discussion_r1848167597)

yummygyudon committed:
[FIX] Response Getter 추가

- API 요청 시, HttpMessageNotWritableException 발생되는 문제 해결

yummygyudon committed:
[FIX] gitignore `*.properties` 추가로 인해 `gradle-wrapper.properties` 삭제되는 문제 해결

yummygyudon committed:
[REFACTOR] Request DTO 내 Json Property 속성 명시

- 클라이언트와의 협업과정에서 명확한 Json 스펙 전달을 위해 고정 Json 키 값을 명시했습니다.

yummygyudon committed:
[FIX] API 메서드 파라미터 중 Body 데이터에 대한 `@RequestBody` 어노테이션 적용

yummygyudon committed:
[REFACTOR] API Response DTO 내 Json Property 속성 명시

yummygyudon committed:
[TEST] Verify API Test

yummygyudon committed:
[TEST] Verify Usecase 테스트

yummygyudon committed:
[REFACTOR] Usecase 결과 - API Response 분리

yummygyudon committed:
[REFACTOR] Usecase 결과에 대한 API Response DTO 변환 메서드 정의 및 반영

yummygyudon committed:
[TEST] Usecase VO 선언 및 Response DTO 분리에 대한 API 테스트 및 Usecase 테스트 변경 반영

yummygyudon committed:
Merge branch 'dev' into feat/#10

yummygyudon committed:
[FEAT] User 도메인 Out Port 추가 - UserRegisterInfo Repository  및 UserRegisterInfo 도메인 정의

- 전화번호 기반 조회 메서드를 선언합니다.

yummygyudon committed:
[FEAT] UserRegisterInfo 데이터 계층(`external`) 객체 구현

- UserRegisterInfoEntity의 경우, UserRegisterInfo 도메인 객체로 변환하는 메서드 정의까지 했습니다.

yummygyudon committed:
[FEAT] User 데이터 계층(`external`) 전화번호 기반 조회 기능 구현

- 데이터가 존재하지 않는 경우에 대한 UserFailure 까지 구현했습니다.

yummygyudon committed:
[MODIFY] 사용자 정보가 존재할 경우에만 인증 이력이 생성되도록 기능 수정

- `REGISTER` 타입일 경우, "동아리 가입자 정보" 데이터 베이스를 조회하도록 합니다.
- `CHANGE`/`SEARCH` 타입의 경우, "기존 가입자 정보" 데이터 베이스를 조회하도록 합니다.

yummygyudon committed:
[MODIFY] 인증 완료 시, 대상자 이름 및 번호 데이터를 함께 반환할 수 있도록 DTO 수정

- 이에 따른 AuthApiControllerTest 코드까지 수정 반영했습니다.

yummygyudon committed:
[MODIFY] 인증 처리 Usecase Return 방식 VO 반환 방식으로의 변경에 따른 반영

- 구현체 VerifyVerificationServiceTest 내 코드 반영까지 완료했습니다.

yummygyudon committed:
[TEST] 인증 생성 UseCase 구현체 Test

Verification Type에 따라 올바른 대상에 대한 조회 기반으로 처리되는지 확인합니다.

sung-silver reviewed (approved):
고생하셨습니다~
